feat(access): Introduce model and API for team-level roles

RyanSkonnord opened the pull request.

RyanSkonnord commented on src/sentry/api/endpoints/organization_member_team_details.py:
```diff
@@ -150,6 +187,43 @@ def post(
+
+                omt.update_team_role(new_role)
+
+        return Response(serialize(team, request.user, TeamWithProjectsSerializer()), status=200)
```
This (`TeamWithProjectsSerializer`) seems like the wrong thing to display at this level of the API. I put it in for the sake of consistency with the POST method. But it's describing the entire team (redundantly to another endpoint?) rather than just the membership object.

Thoughts on creating a new serializer for `OrganizationMemberTeamDetailsEndpoint` and on whether we want to change existing behavior on POST?

markstory replied:
Having the resource be consistent with other methods feels like the right thing to do to me. This `PUT` method could operate on other attributes in the future and reflecting back the entity that was changed is likely what we'd want.

Do we need the `projects` in this response? I thought we were trying to move away from automatically including 1:N relations.

RyanSkonnord replied:
> Having the resource be consistent with other methods feels like the right thing to do to me. This `PUT` method could operate on other attributes in the future and reflecting back the entity that was changed is likely what we'd want.

👍

> Do we need the `projects` in this response? I thought we were trying to move away from automatically including 1:N relations.

I'll look into whether it's safe to take `projects` out of the POST response also.

RyanSkonnord committed:
feat(access): Introduce model and API for team-level roles

wedamija reviewed (commented):
This pr is pretty large, could you split the migration changes out separately?

RyanSkonnord committed:
Check scopes given by team role in has_project_scope

TODO: Add coverage in test_access (both for this and has_team_scope)

RyanSkonnord committed:
Debug has_project_scope; add test_has_scope_from_team_role

RyanSkonnord committed:
Add optional role to create_team_membership

RyanSkonnord committed:
Style tweak

markstory commented on src/sentry/api/endpoints/organization_member_team_details.py:
```diff
@@ -150,6 +187,43 @@ def post(
+            return Response(status=400)
+        result = serializer.validated_data
+
+        if "role" in result:
+            if features.has("organizations:team-roles", organization):
```
Could combine these to save some nesting and make cleanup a 1 line change instead of > 1 lines.

RyanSkonnord replied:
Sure. I want to say I've seen a soft convention where a feature flag check is in its own `if` statement, but I don't mind changing it.

RyanSkonnord commented:
> This pr is pretty large, could you split the migration changes out separately?

I was hesitant to do so at first in case I changed my mind about the model change, but at this point it seems fine.

RyanSkonnord commented on src/sentry/roles/manager.py:
```diff
@@ -31,55 +61,147 @@ def has_scope(self, scope: str) -> bool:
+    def can_manage(self, other: "OrganizationRole") -> bool:
+        return self.priority >= other.priority
+
+    def get_entry_role(self) -> "TeamRole":
```
Will do.

RyanSkonnord commented on src/sentry/models/organizationmember.py:
```diff
@@ -472,19 +457,43 @@ def get_allowed_roles_to_invite(self):
             .exists()
         )
+
+    def update_role(self, role: str) -> None:
```
That's fine. My (admittedly not Pythonic) instinct is to have the role attribute be effectively private, and to have a model method enforce the logic for changing it. But there's no way to enforce that other than convention and I doubt we'll be altering the attribute anywhere else, so moving the code to where it's called should be fine.

dcramer replied:
yeah fwiw i would just prefix them with `_` if you want to communicate that expectation. doesnt need to be an instance method at that piont - anywhere i would see `def _foo()` I would assume its private to local space only and likely a hairy API that you shouldn't reuse without understanding

RyanSkonnord replied:
Is it normal for us to do that with model fields, though? That is,
```python
    _role = models.CharField(max_length=32, null=True, blank=True)
```
Because my wish would be to prevent those who don't know better from directly modifying the field as in `member.role = 'admin'` or `member.update(role='admin')` and instead to guide them toward `member.update_role('admin')` as the public method.

As far as I can find, we don't use underscore-prefixed field names in any models (and it would propagate into SQL if we did, looking odd in Redash, etc.).

dcramer replied:
ah sorry i thought you meant the helper methods

you can technically do that.. id say its not typical and im sure theres lots of strong opinions. like ive definitely setup model fields that are "private" and have getter/setter behavior to ensure its normalized

i think the 'more correct' (albiet more code) way to do that these days is to create a custom django model field that does the validation/serialization/deserialization for you

RyanSkonnord replied:
> i think the 'more correct' (albiet more code) way to do that these days is to create a custom django model field that does the validation/serialization/deserialization for you

Hmm, thanks for the tip. I'll look into that.

RyanSkonnord committed:
Combine feature flag check into `if`

RyanSkonnord committed:
Move "update role" methods from models to endpoints

RyanSkonnord committed:
Add docstring defining entry role

RyanSkonnord committed:
Add extra feature flag check to get_team_role

Generally the flag would need to have been on for a member to have a
non-null role in the first place, but this could make a difference for
an org where the flag has been switched off again.

RyanSkonnord committed:
Use `from __future__ import annotations`

markstory commented on src/sentry/api/endpoints/organization_member_team_details.py:
```diff
@@ -150,6 +187,58 @@ def post(
+
+        serializer = OrganizationMemberTeamSerializer(data=request.data, partial=True)
+        if not serializer.is_valid():
+            return Response(status=400)
```
Should this include validation errors? Right now there are no clues for the API consumer.

RyanSkonnord committed:
Fix comment typo

RyanSkonnord committed:
Merge branch 'master' into team-roles

RyanSkonnord committed:
Merge branch 'master' into team-roles

RyanSkonnord committed:
Add OrganizationGlobalMembership.has_team_membership

Syncing OrganizationGlobalMembership (newly merged from master) with
new has_team_membership method.

RyanSkonnord committed:
Adjust expected_queries in OrganizationDetailsTest

RyanSkonnord committed:
Merge branch 'master' into team-roles

JoshFerge commented on src/sentry/api/endpoints/organization_member_team_details.py:
```diff
@@ -150,6 +187,58 @@ def post(
+                return Response(status=400)
+
+            if not self._can_set_team_role(request, team, new_role):
+                return Response({"detail": ERR_INSUFFICIENT_ROLE}, status=400)
```
should this one be a `403`?

JoshFerge commented on src/sentry/api/endpoints/organization_member_team_details.py:
```diff
@@ -150,6 +187,58 @@ def post(
+        return Response(serialize(team, request.user, TeamWithProjectsSerializer()), status=200)
+
+    @staticmethod
+    def _change_team_member_role(member: OrganizationMemberTeam, role: TeamRole) -> None:
```
nit: could we name this parameter `team_role`?

RyanSkonnord replied:
Done.

JoshFerge commented on src/sentry/api/endpoints/organization_member_team_details.py:
```diff
@@ -150,6 +187,58 @@ def post(
+        if role.priority > entry_role.priority:
+            member.update(role=role.id)
+        else:
+            member.update(role=None)
```
i'm very confused about this case. what are the steps of user action taken to have the omt.role be `None`?

RyanSkonnord replied:
I'd hoped the docstring would explain it; 95fb0344c76055bbaa036cb2b43aca10981f30b7 is my second pass at an inline explanation.

The key other side of this interaction is the following method on the `OrganizationMemberTeam` model: 
```python
    def get_team_role(self) -> TeamRole:
        entry_role = roles.get_entry_role(self.organizationmember.role)
        if self.role:
            team_role = team_roles.get(self.role)
            if team_role.priority > entry_role.priority:
                return team_role
        return entry_role
```
Suppose we have an Org Owner, meaning they are entitled to have a minimum team role of Team Admin on every team. Suppose, for whatever reason, they have a bogus `"contributor"` string in their `OrganizationMemberTeam` object. That `if team_role.priority > entry_role.priority` line ensures that we return the Team Admin role to which they're entitled, and correctly ignore the data calling them a Team Contributor.

A lesser problem is when the persistent role is equal to the entry role. Depending on the path they took to get there, we could have some Org Owners and Managers running around with `"admin"` in their `OrganizationMemberTeam` row and others with null. The outcome would be the same but it would be confusing in Redash, etc. There would also be weird corner cases in case one of those users is demoted: a Manager with a persistent `"admin"` string would remain a Team Admin even after being busted down to Org Member, but one with a null string would not. This is bad because the difference would be opaque until it actually happened.

The point of writing null here is essentially: avoid writing an equal or lesser role in the first place so that the database better reflects reality. Except for the "what happens after they're demoted" case, we don't *need* to do this as long as `get_team_role` is checking `if team_role.priority > entry_role.priority`. We also might change our minds, once we have end-to-end demos to play with, whether it's useful to allow team roles that are explicitly lower than entry roles (e.g., someone is an Org Owner but wants to be designated as a Team Contributor, not Admin, because it better reflects their practical responsibilities).

This logic seems to be incurring a bit of a complexity cost, so I'm open to scrapping it in favor of a simpler implementation, if the above either doesn't make sense or doesn't seem to be worth it. If we do, I believe it's important to communicate in the UI design: "here is the team role that you get by virtue of your org role, and here is the team role that you've been manually assigned, be it higher, lower, or equal."

JoshFerge commented on src/sentry/api/serializers/models/project.py:
```diff
@@ -66,6 +68,17 @@
 
+def _get_team_memberships(team_list: Sequence[Team], user: User) -> Iterable[int]:
+    """Get memberships the user has in the provided team list"""
+    if not user.is_authenticated:
```
when would this case ever happen?

RyanSkonnord replied:
That's probably vestigial from when this was a public method. I'll poke around a little and see if we can delete it.

JoshFerge commented on src/sentry/api/serializers/models/project.py:
```diff
@@ -66,6 +68,17 @@
 LATEST_DEPLOYS_KEY = "latestDeploys"
 
 
+def _get_team_memberships(team_list: Sequence[Team], user: User) -> Iterable[int]:
```
slightly unrelated, but kind of weird we're using `user` here and not `member`.

RyanSkonnord replied:
Er, possibly because the `team_list` could span teams on multiple organizations, therefore there wouldn't be only one relevant `OrganizationMember`? (I don't know for sure if that's possible, though.)

JoshFerge commented on src/sentry/api/serializers/models/team.py:
```diff
@@ -154,7 +158,14 @@ def get_attrs(
 
         for team in item_list:
-            is_member = team.id in memberships
+            membership: OrganizationMemberTeam = memberships.get(team.id)
```
again, this tripped me up -- `team_membership` would be preferred.

RyanSkonnord replied:
Done.

JoshFerge commented on src/sentry/conf/server.py:
```diff
@@ -1625,6 +1639,56 @@ def create_partitioned_queues(name):
+            "event:admin",
+            "project:releases",
+            "project:read",
+            "org:read",
```
not sure if related, but `org:read` seems like a weird scope in general.

JoshFerge commented on tests/sentry/roles/test_manager.py:
```diff
@@ -0,0 +1,147 @@
+        self._assert_entry_is(default_manager, "manager", "admin")
+        self._assert_entry_is(default_manager, "owner", "admin")
+
+    TEST_ORG_ROLES = [
+        {"id": "peasant", "name": "Peasant"},
+        {"id": "baron", "name": "Baron"},
+        {"id": "earl", "name": "Earl"},
+        {"id": "duke", "name": "Duke"},
+        {"id": "monarch", "name": "Monarch"},
+    ]
+
+    TEST_TEAM_ROLES = [
+        {"id": "private", "name": "Private", "is_entry_role_for": "peasant"},
+        {"id": "sergeant", "name": "Sergeant", "is_entry_role_for": "earl"},
+        {"id": "lieutenant", "name": "Lieutenant"},
+        {"id": "captain", "name": "Captain", "is_entry_role_for": "monarch"},
+    ]
```
do these roles test behavior that isn't introduced by our config'd roles? if so, what do you think about just using our roles? i am slow to understand the tests with the medieval names :P

RyanSkonnord replied:
I do like having a set of tests that exercise `RoleManager` independently of what's in `sentry/conf/server.py`. (It's also important to test the actual config'd values, which I think `test_default_manager` does adequately.)

The other reason I really wanted these fake roles is that I needed at least three team roles to fully cover the code (in case of non-default configs). Having Team Admin be the highest team role left a few paths untested, around things being accessible by default to the org "top dog", and so on.

JoshFerge reviewed (commented):
overall architecture looks good to me. small comments on readability. I agree w/ dan, probably worth breaking up into small chunks to merge / give a final 👍🏼  on .

RyanSkonnord committed:
Clarify variable names

RyanSkonnord committed:
Move comment in _change_team_member_role for clarity

RyanSkonnord committed:
Merge branch 'master' into team-roles

RyanSkonnord committed:
Regenerate migration after merge

wedamija commented:
> This pr is pretty large, could you split the migration changes out separately?

I'd still like to see the migration in a separate pr, if it's ready to go.

RyanSkonnord commented:
Thanks for the thorough reviews, all. Closing to break up into small PRs.
